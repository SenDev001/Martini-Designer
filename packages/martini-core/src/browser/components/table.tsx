import { ElementExt } from '@phosphor/domutils';
import { isEqual, pull } from "lodash";
import * as React from "react";
import { Cell, Column, ColumnGroup, ColumnInstance, ColumnInterfaceBasedOnValue, TableCellProps, TableHeaderProps, useFlexLayout, useResizeColumns, UseResizeColumnsColumnOptions, UseResizeColumnsColumnProps, useTable } from "react-table";
import styled from "styled-components";

const Styles = styled.div`
    display: block;
    overflow-x: auto;
    width: 100%;
    height: 100%;
    overflow-y: hidden;

    .table {
        border-spacing: 0;
        display: flex;
        flex-direction: column;
        height: 100%;
        overflow-y: overlay;

        .thead {
            flex-shrink: 0;
        }

        .tbody {
            flex: 1;
            overflow-y: auto;
        }

        .tr {
            :last-child {
                .td {
                    border-bottom: 0;
                }
            }
            .selected {
                background-color: var(--theia-list-activeSelectionBackground);
            }
        }

        .th {
            user-select: none;
            padding: 0.5rem;
        }

        .td {
            min-height: 25px;
        }

        .th,
        .td {
            margin: 0;
            border-right: 1px solid var(--martini-table-border);
            overflow-x: hidden;
            position: relative;

            :last-child {
                border-right: 0;
            }

            .resizer {
                right: 0;
                width: 10px;
                height: 100%;
                position: absolute;
                top: 0;
                z-index: 1;
                touch-action :none;
            }

            .resizer:hover {
                background-color: var(--theia-scrollbarSlider-background);
            }
        }
    }
`;

export type TableColumn = Column<any> & UseResizeColumnsColumnOptions<any> & ColumnInterfaceBasedOnValue<any> & {
    /**
     * Cell padding.
     */
    cellPadding?: string;
};

export interface CellClickEvent {
    columnId: string;
    rowIndex: number;
    value: any;
    mouseEvent: React.MouseEvent<HTMLDivElement>;
}

export interface TableProps {
    columns: TableColumn[];
    data: any[];
    /**
     * Defaults to true.
     */
    showTableHeader?: boolean;
    focusable?: boolean;
    onCellClick?: (e: CellClickEvent) => void;
    onCellDoubleClick?: (e: CellClickEvent) => void;
    /**
     * Defaults to true.
     */
    rowSelectionEnabled?: boolean;
    /**
     * Defaults to true.
     */
    multiSelection?: boolean;
    selectedRows?: number[];
    onSelectionChange?: (selectedRows: number[]) => void;
    tableRef?: React.MutableRefObject<any>;
}

export const Table: React.FC<TableProps> = ({
    columns,
    data,
    showTableHeader = true,
    focusable,
    onCellClick,
    onCellDoubleClick,
    rowSelectionEnabled = true,
    multiSelection = true,
    selectedRows: selectedRowsProp,
    onSelectionChange,
    tableRef
}) => {
    const defaultColumn = React.useMemo(
        () => ({
            minWidth: 30,
            width: 100,
            maxWidth: 400
        } as Partial<TableColumn>),
        []
    );
    const headerProps = (props: TableHeaderProps, { column }: any) => getStyles(props, column.align);

    const cellProps = (props: TableCellProps, { cell }: any) => getStyles(props, cell.column.align, cell.column.cellPadding);

    const getStyles = (props: TableHeaderProps | TableCellProps, align = "left", padding?: string) => [
        props,
        {
            style: {
                justifyContent: align === "right" ? "flex-end" : "flex-start",
                alignItems: "center",
                display: "flex",
                overflow: "visible",
                padding
            }
        }
    ];

    const {
        getTableProps,
        headerGroups,
        rows,
        prepareRow
    } = useTable<any>({
        columns,
        data,
        defaultColumn
    }, useFlexLayout, useResizeColumns);

    const [selectedRows, setSelectedRows] = React.useState<number[]>(selectedRowsProp || []);
    const selectedRowRef = React.useRef<HTMLDivElement>(null);
    const lastMouseSelectedRow = React.useRef<number>();

    React.useEffect(() => setSelectedRows(selectedRowsProp || []), [selectedRowsProp]);

    React.useEffect(() => {
        if (selectedRowRef.current)
            ElementExt.scrollIntoViewIfNeeded(selectedRowRef.current.parentElement!, selectedRowRef.current);
    }, [selectedRows]);

    const handleSelectionChange = (selectedRows: number[]) => {
        if (onSelectionChange)
            onSelectionChange(selectedRows);
        if (!selectedRowsProp)
            setSelectedRows(selectedRows);
    };

    const handleCellClick = (e: React.MouseEvent<HTMLDivElement>, cell: Cell<any>) => {
        if (rowSelectionEnabled) {
            const _selectedRows = [...selectedRows];

            if (e.metaKey && multiSelection) {
                if (_selectedRows.includes(cell.row.index)) {
                    pull(_selectedRows, cell.row.index);
                    lastMouseSelectedRow.current = _selectedRows[_selectedRows.length - 1];
                }
                else {
                    _selectedRows.push(cell.row.index);
                    lastMouseSelectedRow.current = cell.row.index;
                }
            }
            else if (e.shiftKey && _selectedRows.length > 0 && multiSelection) {
                lastMouseSelectedRow.current = cell.row.index;
                const lastSelected = _selectedRows[_selectedRows.length - 1];
                e.preventDefault();
                if (document.getSelection)
                    document.getSelection()?.removeAllRanges();

                if (lastSelected < cell.row.index) {
                    for (let i = lastSelected + 1; i <= cell.row.index; i++) {
                        if (_selectedRows.includes(i))
                            pull(_selectedRows, i);
                        else
                            _selectedRows.push(i);
                    }
                }
                else if (lastSelected > cell.row.index) {
                    for (let i = lastSelected - 1; i >= cell.row.index; i--) {
                        if (_selectedRows.includes(i))
                            pull(_selectedRows, i);
                        else
                            _selectedRows.push(i);
                    }
                }
            }
            else {
                _selectedRows.splice(0, _selectedRows.length);
                _selectedRows.push(cell.row.index);
                lastMouseSelectedRow.current = cell.row.index;
            }

            if (!isEqual(selectedRows, _selectedRows))
                handleSelectionChange(_selectedRows);
        }
        if (onCellClick) {
            onCellClick({
                columnId: cell.column.id,
                rowIndex: cell.row.index,
                value: cell.row.original,
                mouseEvent: e
            });
        }
    };

    const handleCellDoubleClick = (e: React.MouseEvent<HTMLDivElement>, cell: Cell<any>) => {
        if (onCellDoubleClick) {
            onCellDoubleClick({
                columnId: cell.column.id,
                rowIndex: cell.row.index,
                value: cell.row.original,
                mouseEvent: e
            });
        }
    };

    const handleKeyDown = (e: React.KeyboardEvent) => {
        if (rowSelectionEnabled && (e.key === "ArrowUp" || e.key === "ArrowDown")) {
            if (e.target instanceof HTMLInputElement)
                return;

            if (e.shiftKey && multiSelection) {
                const _selectedRows = [...selectedRows];
                const first = lastMouseSelectedRow.current || _selectedRows[0];
                const last = _selectedRows[_selectedRows.length - 1];
                if (e.key === "ArrowUp") {
                    let next = last - 1;
                    if (last <= first) {
                        while (next >= 0 && _selectedRows.includes(next)) {
                            pull(_selectedRows, next);
                            _selectedRows.push(next);
                            next--;
                        }
                        if (next >= 0)
                            _selectedRows.push(next);
                    } else if (last > first)
                        pull(_selectedRows, last);
                } else {
                    let next = last + 1;
                    if (last >= first && next < data.length) {
                        while (next < data.length && _selectedRows.includes(next)) {
                            pull(_selectedRows, next);
                            _selectedRows.push(next);
                            next++;
                        }

                        if (next < data.length)
                            _selectedRows.push(next);
                    } else if (last < first)
                        pull(_selectedRows, last);
                }

                handleSelectionChange(_selectedRows);
            } else {
                let index = selectedRows.length > 0 ? selectedRows[0] : -1;

                if (e.key === "ArrowUp")
                    index--;
                else
                    index++;

                if (index < 0)
                    index = data.length - 1;
                else if (index >= data.length)
                    index = 0;

                handleSelectionChange([index]);
            }

            e.preventDefault();
        }
    };

    return (
        <Styles
            tabIndex={focusable ? 0 : undefined}
            onKeyDown={handleKeyDown}
            ref={tableRef}
        >
            <div {...getTableProps()} className="table">
                {showTableHeader && <div className="thead">
                    {headerGroups.map(headerGroup => (
                        <div {...headerGroup.getHeaderGroupProps()} className="tr">
                            {headerGroup.headers.map(
                                // @ts-ignore
                                (column: ColumnInstance & ColumnGroup<any> & UseResizeColumnsColumnProps<any>) => (
                                    <div {...column.getHeaderProps(headerProps)} className="th">
                                        {column.render("Header")}
                                        {column.canResize && column.accessor !== undefined && (
                                            <div
                                                {...column.getResizerProps()}
                                                className="resizer"
                                            />
                                        )}
                                    </div>
                                ))}
                        </div>))}
                </div>}
                <div className="tbody">
                    {rows.map(row => {
                        prepareRow(row);

                        let className = "td";

                        if (selectedRows.includes(row.index))
                            className += " selected";

                        return (
                            <div {...row.getRowProps()} className="tr" ref={row.index === selectedRows[selectedRows.length - 1] ? selectedRowRef : undefined}>
                                {row.cells.map(cell => (
                                    <div {...cell.getCellProps(cellProps)}
                                        className={className}
                                        onClick={e => handleCellClick(e, cell)}
                                        onDoubleClick={e => handleCellDoubleClick(e, cell)}
                                        onContextMenu={e => handleCellClick(e, cell)}
                                    >
                                        {cell.render("Cell")}
                                    </div>
                                ))}
                            </div>
                        );
                    })}
                </div>
            </div>
        </Styles>
    );
};
