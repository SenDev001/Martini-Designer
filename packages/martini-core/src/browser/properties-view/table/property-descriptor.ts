import { Event } from "@theia/core";
import { CellEditorProvider } from "../../components/cell-editors";

export interface PropertyDescriptor {
    target: any;
    name: string;
    displayName: string;
    description: string;
    readOnly: boolean;
    value: any;
    /**
     * Used to reset the property.
     */
    defaultValue?: any;
    /**
     * Used to determine if the reset menu should be shown.
     */
    hasDefaultValue?: boolean;
    validate?: (value: any) => string | undefined;
    getCellEditor?: CellEditorProvider<any>;
}

export interface PropertySource {
    getPropertyDescriptors(): PropertyDescriptor[];

    setProperty(name: string, newValue: any): void;

    iconClass?: string;
    title?: string;
    onPropertyChange?: Event<void>;

    dispose?(): void;
}

export interface PropertySourceFactory {
    support(selection: object): boolean;

    getPropertySource(selection: object): PropertySource;
}

export const PropertySourceFactory = Symbol("PropertySourceFactory");
