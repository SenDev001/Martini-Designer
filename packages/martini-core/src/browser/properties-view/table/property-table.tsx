import { CommandRegistry } from "@phosphor/commands";
import { Menu } from "@phosphor/widgets";
import { toString } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import styled from "styled-components";
import { CellClickEvent, Table, TableColumn } from "../../components/table";
import { PropertyDescriptor } from "./property-descriptor";

export interface PropertiesTableProps {
    descriptors: PropertyDescriptor[];
    onApply: (descriptor: PropertyDescriptor, newValue: any) => void;
}

const PropertyTableStyles = styled.div`
    overflow-y: hidden;
    height: 100%;

    .table {
        .td {
            border-bottom: 1px solid var(--martini-table-border);
        }
    }
`;

const PropertyNameCell = styled.div`
    display: grid;
    grid-template-columns: 1fr max-content;
    width: 100%;

    .reset {
        cursor: pointer;
        visibility: hidden;
    }

    :hover .reset {
        visibility: visible;
    }
`;

export const PropertyTable: React.FC<PropertiesTableProps> = ({
    descriptors,
    onApply
}) => {
    const reset = (descriptor: PropertyDescriptor) => {
        if (onApply && !descriptor.readOnly)
            onApply(descriptor, descriptor.defaultValue);
    };

    const handleCellClick = (e: CellClickEvent) => {
        const descriptor = e.value as PropertyDescriptor;
        // only show reset menu for the property name column
        if (!descriptor || !descriptor.hasDefaultValue || descriptor.readOnly ||
            e.columnId !== "displayName" || e.mouseEvent.button !== 2)
            return;

        const commands = new CommandRegistry();
        commands.addCommand("reset", {
            label: messages.reset_default,
            execute: () => reset(descriptor)
        });

        const menu = new Menu({
            commands
        });
        menu.addItem({
            command: "reset"
        });
        menu.open(e.mouseEvent.clientX, e.mouseEvent.clientY);
        e.mouseEvent.preventDefault();
    };

    const columns: TableColumn[] = [
        {
            Header: "Property",
            accessor: "displayName",
            cellPadding: "0 0 0 4px",
            Cell: props => {
                const descriptor = props.row.original as PropertyDescriptor;
                return <PropertyNameCell
                    title={descriptor.description}
                >
                    {descriptor.displayName}
                    {(!descriptor.readOnly && (descriptor.hasDefaultValue || descriptor.defaultValue !== undefined)) && <div
                        title={messages.reset_default}
                        className="reset martini-tree-icon martini-reset-icon"
                        onClick={() => reset(descriptor)}
                    />}
                </PropertyNameCell>;
            }
        },
        {
            Header: "Value",
            Cell: props => {
                const descriptor = props.data[props.row.index] as PropertyDescriptor;

                return !descriptor.getCellEditor ?
                    <div style={{ paddingLeft: "4px" }}>{toString(descriptor.value)}</div> :
                    descriptor.getCellEditor({
                        editor: {
                            get value() { return descriptor.value; },
                            onApply: value => {
                                if (!descriptor.readOnly)
                                    onApply(descriptor, value);
                            },
                            validate: descriptor.validate,
                            readOnly: descriptor.readOnly
                        }
                    });
            }
        }
    ] as TableColumn[];

    return <PropertyTableStyles>
        <Table
            columns={columns}
            data={descriptors}
            onCellClick={handleCellClick}
            rowSelectionEnabled={false}
        />
    </PropertyTableStyles>;
};
