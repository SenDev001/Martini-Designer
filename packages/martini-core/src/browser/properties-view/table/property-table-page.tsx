import { Widget } from "@phosphor/widgets";
import { injectable, multiInject } from "inversify";
import * as React from "react";
import { AbstractPropertiesViewPage } from "../properties-view-page";
import { PropertyDescriptor, PropertySource, PropertySourceFactory } from "./property-descriptor";
import { PropertyTable } from "./property-table";

/**
 * Base class for properties view page with a property table.
 */
@injectable()
export abstract class AbstractPropertyTablePage extends AbstractPropertiesViewPage {
    @multiInject(PropertySourceFactory)
    private readonly factories: PropertySourceFactory[];
    private propertySource: PropertySource | undefined;
    scrollY = false;

    onSelectionChange(source: Widget, selection: object): void {
        if (this.propertySource && this.propertySource.dispose)
            this.propertySource.dispose();

        if (selection) {
            this.propertySource = this.factories
                .find(p => p.support(selection))
                ?.getPropertySource(selection);
        }

        if (this.propertySource && this.propertySource.onPropertyChange)
            this.propertySource.onPropertyChange(() => this.onChangeEmitter.fire());
    }

    protected doRender(): React.ReactNode {
        if (this.propertySource) {
            const descriptors: PropertyDescriptor[] = this.propertySource.getPropertyDescriptors();
            const onApply = (descriptor: PropertyDescriptor, value: any) =>
                this.applyValue(descriptor, value);
            return (
                <PropertyTable
                    descriptors={descriptors}
                    onApply={onApply}
                />
            );
        } else return undefined;
    }

    private applyValue(descriptor: PropertyDescriptor, newValue: any) {
        this.propertySource?.setProperty(descriptor.name, newValue);
    }
}
