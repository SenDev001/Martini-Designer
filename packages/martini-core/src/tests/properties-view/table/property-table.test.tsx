import * as React from "react";
import * as Renderer from "react-test-renderer";
import { PropertyTable } from "../../../browser/properties-view/table/property-table";
import { PropertyDescriptor } from "../../../browser/properties-view/table/property-descriptor";

const descriptors: PropertyDescriptor[] = [
    {
        name: "firstName",
        displayName: "First Name",
        description: "It's the first name, duh!",
        value: "John",
        defaultValue: "Jean",
        hasDefaultValue: true,
        target: {},
        readOnly: false
    },
    {
        name: "lastName",
        displayName: "Last Name",
        description: "C'mon, it's the last name, dude...",
        value: "Doe",
        target: {},
        readOnly: false
    },
    {
        name: "age",
        displayName: "Age",
        description: "C'est l'age, idiot.",
        value: 30,
        target: {},
        readOnly: true
    }
];

test("PropertyTable should be rendered with the descriptors", () => {
    const component = Renderer.create(
        <PropertyTable descriptors={descriptors} onApply={() => { }} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("PropertyTable should be rendered without descriptors", () => {
    const component = Renderer.create(
        <PropertyTable descriptors={[]} onApply={() => { }} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
