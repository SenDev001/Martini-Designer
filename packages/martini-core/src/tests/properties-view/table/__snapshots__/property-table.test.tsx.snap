// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyTable should be rendered with the descriptors 1`] = `
<div
  className="sc-AxirZ kdLRCK"
>
  <div
    className="sc-AxjAm ljDiNr"
    onKeyDown={[Function]}
  >
    <div
      className="table"
      role="table"
      style={
        Object {
          "minWidth": "200px",
        }
      }
    >
      <div
        className="thead"
      >
        <div
          className="tr"
          role="row"
          style={
            Object {
              "display": "flex",
              "flex": "1 0 auto",
              "minWidth": "60px",
            }
          }
        >
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "position": "relative",
                "width": "100px",
              }
            }
          >
            Property
            <div
              className="resizer"
              draggable={false}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              role="separator"
              style={
                Object {
                  "cursor": "ew-resize",
                }
              }
            />
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "position": "relative",
                "width": "100px",
              }
            }
          >
            Value
          </div>
        </div>
      </div>
      <div
        className="tbody"
      >
        <div
          className="tr"
          role="row"
          style={
            Object {
              "display": "flex",
              "flex": "1 0 auto",
              "minWidth": "60px",
            }
          }
        >
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": "0 0 0 4px",
                "width": "100px",
              }
            }
          >
            <div
              className="sc-AxiKw cSXCKF"
              title="It's the first name, duh!"
            >
              First Name
              <div
                className="reset martini-tree-icon martini-reset-icon"
                onClick={[Function]}
                title="Reset to Default"
              />
            </div>
          </div>
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "width": "100px",
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "4px",
                }
              }
            >
              John
            </div>
          </div>
        </div>
        <div
          className="tr"
          role="row"
          style={
            Object {
              "display": "flex",
              "flex": "1 0 auto",
              "minWidth": "60px",
            }
          }
        >
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": "0 0 0 4px",
                "width": "100px",
              }
            }
          >
            <div
              className="sc-AxiKw cSXCKF"
              title="C'mon, it's the last name, dude..."
            >
              Last Name
            </div>
          </div>
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "width": "100px",
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "4px",
                }
              }
            >
              Doe
            </div>
          </div>
        </div>
        <div
          className="tr"
          role="row"
          style={
            Object {
              "display": "flex",
              "flex": "1 0 auto",
              "minWidth": "60px",
            }
          }
        >
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": "0 0 0 4px",
                "width": "100px",
              }
            }
          >
            <div
              className="sc-AxiKw cSXCKF"
              title="C'est l'age, idiot."
            >
              Age
            </div>
          </div>
          <div
            className="td"
            onClick={[Function]}
            onContextMenu={[Function]}
            onDoubleClick={[Function]}
            role="cell"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "width": "100px",
              }
            }
          >
            <div
              style={
                Object {
                  "paddingLeft": "4px",
                }
              }
            >
              30
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PropertyTable should be rendered without descriptors 1`] = `
<div
  className="sc-AxirZ kdLRCK"
>
  <div
    className="sc-AxjAm ljDiNr"
    onKeyDown={[Function]}
  >
    <div
      className="table"
      role="table"
      style={
        Object {
          "minWidth": "200px",
        }
      }
    >
      <div
        className="thead"
      >
        <div
          className="tr"
          role="row"
          style={
            Object {
              "display": "flex",
              "flex": "1 0 auto",
              "minWidth": "60px",
            }
          }
        >
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "position": "relative",
                "width": "100px",
              }
            }
          >
            Property
            <div
              className="resizer"
              draggable={false}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              role="separator"
              style={
                Object {
                  "cursor": "ew-resize",
                }
              }
            />
          </div>
          <div
            className="th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "alignItems": "center",
                "boxSizing": "border-box",
                "display": "flex",
                "flex": "100 0 auto",
                "justifyContent": "flex-start",
                "minWidth": "30px",
                "overflow": "visible",
                "padding": undefined,
                "position": "relative",
                "width": "100px",
              }
            }
          >
            Value
          </div>
        </div>
      </div>
      <div
        className="tbody"
      />
    </div>
  </div>
</div>
`;
